upstairs/src/lib.rs

```diff
@@ -3123,7 +3085,36 @@ impl Downstairs {
         assert_eq!(old_state, IOState::New);
         self.io_state_count.decr(&old_state, client_id);
         self.io_state_count.incr(&new_state, client_id);
-        Some(job.work.clone())
+        let mut out = job.work.clone();
+        drop(handle);
+
+        if self.dependencies_need_cleanup(client_id) {
+            match &mut out {
```

Comment:
We could almost put the match as part of the `dependencies_need_cleanup()` check, as
really that logic should live all in one place, but you have improved what was here before
enough already that we don't need to do that.